DDS-872-eventual_consistency_status

dmlond opened the pull request:
attempt to push smaller commits for review
more to come

dmlond committed:
Merge pull request #960 from Duke-Translational-Bioinformatics/develop

Merge v1.2.6, v1.2.7, v1.2.8 to ua_test

dmann committed:
Initial ErrorQueueHandler service interface

dmann committed:
Configuring BunnyMock to use BunnyQueue#pop api

dmann committed:
Test enqueueing mocked messages to error queue

dmann committed:
ErrorQueueHandler#message_count returns the number of messages in error queue

dmann committed:
Refactor ErrorQueueHandler to use single method to connect to error_queue

dmann committed:
Allow routing_key to be set in enqueue_mocked_message method

dmann committed:
Adding nack functionality to BunnyMock

dmann committed:
Initial implementation of ErrorQueueHandler#messages, ignoring arguments/keywords

dmann committed:
ErrorQueueHandler#messages allows routing_key scoping

dmann committed:
ErrorQueueHandler#messages allows limit

dmann committed:
Ensure :routing_key and :limit play nice together

dmann committed:
Refactor ErrorQueueHandler message serialization

dmann committed:
Initial implementation of ErrorQueueHandler#requeue_message

dmann committed:
Initial implementation of ErrorQueueHandler#requeue_all

dmann committed:
Initial implementation of ErrorQueueHandler#requeue_messages

dmann committed:
ErrorQueueHandler#requeue_all sends nack when Bunny::Exception raised

dmann committed:
Refactor methods to use #republish_message

dmann committed:
begin;ensure is a bit cleaner than begin;rescue;raise

dmann committed:
RabbitMQ does not like when you nack an acked delivery tag...

dmann committed:
ErrorQueueHandler#requeue_message sends nack when Bunny::Exception raised

dmann committed:
A shared_context for Bunny::Exception testing

dmann committed:
ErrorQueueHandler#requeue_messages sends nack when Bunny::Exception raised

dmann committed:
Unify method implementations before refactoring

dmann committed:
Testing #requeue_message with last message id

dmann committed:
Refactor message iteration into #each_error_queue_message

dmann committed:
Shared context for error queue message utilities

dmann committed:
Initial rake tasks for queue:errors:message_count and :messages

dmann committed:
Remove unneeded comments

dmann committed:
A shared_context for setting ENV var before and unsetting after examples

dmann committed:
Pass ROUTING_KEY and LIMIT env to ErrorQueueHandler#messages

dmann committed:
Initiial implementation of queue:errors:requeue_message rake task

dmann committed:
queue:errors:requeue_message properly handles non-existent messages and raised exceptions

dmann committed:
Initiial implementation of queue:errors:requeue_all rake task

dmann committed:
Initiial implementation of queue:errors:requeue_messages rake task

dmann committed:
Do not need to test presence of required routing_key

dmann committed:
Merge branch 'develop' into DDS-869-error_queue_handler

dmlond committed:
Merge pull request #984 from dmann/DDS-869-error_queue_handler

DDS-869 error queue handler

dmann committed:
Merge remote-tracking branch 'upstream/ua_test' into develop

dmlond committed:
Merge pull request #991 from dmann/develop

Updating develop with ua_test

dmlond committed:
deleted relations should not cause validation failures

dmann committed:
Merge pull request #992 from dmlond/DDS-874_prov_relation_bugfix

deleted relations should not cause validation failures

dmann committed:
Merge pull request #990 from Duke-Translational-Bioinformatics/develop

Merge background jobs (v1.3) to ua_test

dmann committed:
Enqueue jobs using after_create, after_update, or after_save callbacks

dmann committed:
Ensure Project is not committed to DB when ::perform_later fails

dmann committed:
Organize method specific tests in describe blocks

dmann committed:
Ensure DataFile (using Elasticsearch::Model) is not committed to DB when ::perform_later fails

dmlond committed:
Merge pull request #995 from dmann/DDS-875-enqueue_background_jobs_after_save

Enqueue jobs using after_create, after_update, or after_save callbacks

dmann committed:
Include all available Sneakers and Bunny configuration in sneakers initializer.

dmann committed:
Merge pull request #996 from dmann/DDS-876-sneakers_configuration_via_env

DDS-876 sneakers configuration via env

dmann committed:
Cast string input to Integer when cromulent

dmann committed:
Merge pull request #997 from dmann/DDS-876-sneakers_configuration_via_env

DDS-786 Cast string input to Integer when cromulent

dmann committed:
Initial sprawl workflow commit

dmann committed:
Recreate job_transactions table with UUID primary key

dmann committed:
Merge pull request #999 from dmann/DDS-878-change_job_transactions_id_to_uuid

DDS-878 Recreate job_transactions table with UUID primary key

dmann committed:
Merge remote-tracking branch 'upstream/ua_test' into develop

dmann committed:
Merge pull request #1001 from dmann/develop

Merge ua_test to develop

dmann committed:
Merge pull request #1000 from Duke-Translational-Bioinformatics/develop

Merge v1.3 bug fixes to ua_test

dmann committed:
Allowing queue existence test to be skipped via ENV

dmann committed:
Merge pull request #1002 from dmann/DDS-880-bunny_channel_close_bug

Allowing queue existence test to be skipped via ENV

dmann committed:
Merge remote-tracking branch 'upstream/develop' into DDS-879-workflow_sprawl

dmann committed:
Child class queue should include all expected arguments

dmann committed:
Storing one class instance of Publisher per queue. Should reduce RabbitMQ channel sprawl.

dmann committed:
Merge remote-tracking branch 'upstream/ua_test' into DDS-880-bunny_channel_close_bug

dmann committed:
Fix typo

dmann committed:
Merge pull request #1004 from dmann/DDS-880-bunny_channel_close_bug

DDS-880 bunny channel close bug

dmann committed:
Merge pull request #1005 from Duke-Translational-Bioinformatics/develop

Background Job bug fix (v1.3.13) to ua_test

dmann committed:
Merge remote-tracking branch 'upstream/develop' into DDS-879-workflow_sprawl

dmann committed:
Merge remote-tracking branch 'upstream/ua_test' into DDS-879-workflow_sprawl

dmann committed:
Mentioning env var requirements in usage output

dmann committed:
Remove commented out code...

dmann committed:
Merge pull request #1007 from dmann/DDS-879-workflow_sprawl

DDS-879 workflow sprawl script

dmann committed:
shared_examples to test elasticsearch indexing

dmann committed:
MessageLogWorker creates an index in elasticsearch

dmann committed:
Elasticsearch document type is set to routing_key of message

dmann committed:
Elasticsearch document index is "queue_messages"

dmann committed:
Indexing payload, delivery_info, and properties as json strings to avoid Elastisearch::Transport errors

dmlond committed:
Merge pull request #1008 from dmann/DDS-877-message_log_worker_elasticsearch_indexing

DDS-877 message log worker elasticsearch indexing

dmann committed:
Reconfigure elasticsearch docker to replicate Bonsai

dmann committed:
Reconfigure elasticsearch docker for circleci to replicate Bonsai

dmann committed:
Ensure indexes exist in elasticsearch when testing

dmann committed:
Merge pull request #1009 from dmann/DDS-877-message_log_worker_elasticsearch_indexing

Reconfigure elasticsearch docker to replicate Bonsai

dmlond committed:
Merge pull request #1010 from Duke-Translational-Bioinformatics/develop

Merge  develop (1.3.14, 1.3.15) to ua_test

dmann committed:
The message_log queue needs a retry error queue separate from other jobs

dmann committed:
Merge pull request #1011 from dmann/DDS-877-message_log_worker_elasticsearch_indexing

DDS-877 message log worker elasticsearch indexing

dmann committed:
Initial attempt at handling ActiveJob::DeserializationError

dmann committed:
Using an instance variable to keep track of retries

Donmurry committed:
Merge pull request #1012 from Duke-Translational-Bioinformatics/develop

Merge develop (v1.3.16) to ua_test

dmann committed:
Wait before retrying execute after DeserializationError

dmann committed:
Store interval for DeserializationError retries

dmann committed:
Configurable wait time for DeserializationError retries

dmann committed:
Config error retry via APPLICATION_JOB_DESERIALIZATION_ERROR_RETRY_INTERVAL env var

Donmurry committed:
Merge pull request #1016 from dmann/DDS-884-handle_argument_deserialization_exception

DDS-884 handle argument deserialization exception

dmann committed:
Merge remote-tracking branch 'upstream/ua_test' into develop

dmann committed:
First pass at MessageLogQueueHandler

dmlond committed:
add_eventual_consistency_support migration:
- projects and uploads now have is_consistent boolean attribute
- uploads also have has_integrity_exception boolean attribute

UploadSpec expects:
- #complete no longer interacts with swift (will enqueue a job in a future commit)

- #make_consistent
  - refactors code that was in complete
  - intended for use in job
  - sets is_consistent to true unless a non IntegrityException Exception occurs
  - calls #integrity_exception for IntegrityExceptions

- #integrity_exception
    - sets has_integrity_exception to true
    - sets error_at
    - sets error_message to the integrity exception reason
    - no longer raises a new IntegrityException

Upload meets spec

dmann committed:
Check delivery_info and meta_data arguments are passed to work_with_params

dmann committed:
Use Faker when stubbing message content_type

dmlond committed:
Merge branch 'DDS-872-eventual_consistency_status' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-872-eventual_consistency_status

dmlond committed:
UploadCompletionJob
  manages process of making an upload consistent

dmlond committed:
UploadSpec expects Upload
  - to raise ConsistencyException if temporary_url is called on an upload that is not consistent
  - to behave_like a job_transactionable model
  - to enqueue an UploadCompletionJob during complete
Upload matches expection
new ConsistencyException for when a model that is not consistent is used in a request

dmlond committed:
ProjectStorageProviderInitializationJobSpec expects ProjectStorageProviderInitializationJob
to set project.is_consistent to true after it initializes the storage_provider

dmlond committed:
ProjectSerializerSpec and UploadSerializerSpec expect their serialized data to
include :is_consistent

UploadSerializerSpec also expects the 'status' to include 'has_integrity_exception'
Models meet expectations

dmlond committed:
ProjectSerializerSpec and UploadSerializerSpec expect their serialized data to
include :is_consistent

UploadSerializerSpec also expects the 'status' to include 'has_integrity_exception'
Models meet expectations

dmlond committed:
Merge branch 'DDS-872-eventual_consistency_status' of github.com:dmlond/duke-data-service into DDS-872-eventual_consistency_status

dmlond committed:
changed status to in progress
added code to example IntegrityException response

dmlond committed:
api_design/DDS--872-eventual_consistency_status.md speling

modified cassettes for app_api_spec queue tests to fix failing request

dmlond committed:
added 'code' to standard error response. Default is 'not_provided'
This will need to be refactored to DRY

dmann committed:
Attribute accessor #work_message_limit for MessageLogWorker

dmlond committed:
requests_shared_examples:
  no endpoint will depend on calls to the storage_provider, so the
'storage_provider backed resource' shared_example has been removed

uploads_api_spec
  does not need vcr for tests of the /complete endpoint
  does not need to prep for, and test integrity exceptions (will be
  made eventually consistent, and this code will end up in
  the files_api and file_versions_api spec tests

uploads_api
  adds 'code' to error response

dmann committed:
A method_sym not required for elasticsearch indexer shared example

dmann committed:
Wrap #index_messages call with elasticsearch indexer shared example to clean up documents

dmlond committed:
- requests_shared_examples added shared_example 'an eventually consistent resource':
    - requires eventually_consistent_resource, a resource that must
      be consistent for the request to succceed
    - updates eventually_consistent_resource to is_consistent: false
    - calls endpoint
    - expects endpoint to return the standard response for eventual consistency
      exceptions

- uploads_api_spec expects POST /projects/:id/uploads to behave like an eventually consistent
  resource for the eventually consistent project

- base.rb
  - adds consistent? helper which raises ConsistencyException for an object
    unless object.is_consistent?
  - adds rescue_from ConsistencyException for the standard ConsistencyException error

- uploads_api meets expectation, adds swagger doc for consistency expectation

dmlond committed:
- uploads_api_spec expects PUT /uploads/:id/complete to respond with 202 Accepted
  for a successful action, to distinguish it as an eventually consistent operation

- uploads_api:
  - no longer includes rescue for IntegrityException (this will be moved
    to V1::Base to be shared by files_api and file_versions_api.
  - satisfies 202 response expectation

dmlond committed:
- projects_api_spec expects POST /projects to return 202 Accepted
  to signal eventual consistency
- projects_api meets expectation

dmlond committed:
- request_shared_examples added shared_example 'an eventually consistent upload integrity exception'
  which requires an upload, and tests that an IntegrityException error is returned
  when the upload tied to a resource has an integrity exception

- files_api_spec expects GET /files/:id/url to
  - behave like an eventually consistent upload integrity exception
  - behave like an an eventually consistent resource

- file_version_api_spec expects GET /file_versions/:id/url to
  - behave like an eventually consistent upload integrity exception
  - behave like an an eventually consistent resource

- files_api and file_versions_api meet expectations

dmlond committed:
base:
  - renamed consistent? to check_consistency! for consistent helper nomenclature
  - added check_integrity! to check if an upload has_integrity_exception and raise IntegrityException

uploads_api renamed consistent? to check_consistency!

file_versions_api and files_api implement check_consistency! and check_integrity!
to throw expected IntegrityExceptions

file_versions_api creates consistent upload for test

dmlond committed:
removed error codes that will no longer be thrown by the endpoint

dmlond commented:
ready for official review :)

dmann commented on app/jobs/project_storage_provider_initialization_job.rb:
```diff
@@ -4,6 +4,7 @@ class ProjectStorageProviderInitializationJob < ApplicationJob
   def perform(job_transaction:, storage_provider:, project:)
     self.class.start_job job_transaction
     storage_provider.put_container(project.id)
+    project.update(is_consistent: true)
```
Should handle a failed update here. Maybe call `update!` instead, which will throw an exception if a validation error occurs.

dmann commented on app/models/upload.rb:
```diff
@@ -52,18 +55,28 @@ def manifest
+    end
+  end
+
+  def make_consistent
```
Would it be possible to have a more descriptive name for this method? `make_consistent` is kind of vague IMHO.

dmlond replied:
ARGH! vcr! :{

dmann commented on app/models/upload.rb:
```diff
@@ -52,18 +55,28 @@ def manifest
+      unless meta["content-length"].to_i == size
+        raise IntegrityException, "reported size does not match size computed by StorageProvider"
       end
+      update_attributes({
```
`update_attributes` is an alias for `update`[1] and will run all the validations and whatnot on the object. I'm not sure what will happen here in the case of a validation failure.

1. http://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-update_attributes

dmann commented on app/models/upload.rb:
```diff
@@ -81,8 +94,8 @@ def integrity_exception(message)
     fingerprints.reload
     update_attributes({
```
Possibly another case for `update!`

dmann commented on app/serializers/upload_serializer.rb:
```diff
@@ -1,16 +1,21 @@
     {
       initiated_on: object.created_at,
       completed_on: object.completed_at,
+      has_integrity_exception: object.has_integrity_exception?,
```
We have a couple issues here. First, the attribute `has_integrity_exception` isn't actually set anywhere, so this will always return `false`. That's fine, because I don't think that we actually need that attribute. The integrity of the upload can be inferred from the presence of the `error_message`, which is set when an integrity exception is raised.

If uploads start using `error_message` for non-integrity related exceptions, I suggest we add a `error_type` attribute.

dmann reviewed (changes requested):
Just a couple changes, nothing major though.

dmann committed:
Use MessageLogWorker#run to handle indexing

dmann committed:
Attribute #work_duration for duration configuration

dmlond committed:
update! used to raise an exception if the project is somehow made invalid
before the job runs, which allows the exception to be caught in the job
error_log

dmlond committed:
- renamed #make_consistent to #create_and_validate_storage_manifest
- required [vcr cassette storage cabinet](http://gordogato.com/oscommerce/catalog/images/zm_22x_vhs_storage_1.jpg) to be re-labelled

dmlond committed:
should have gone with the last commit

dmlond committed:
uses upload! instead of upload_attributes
  - upload_attributes is an alias for upload
  - upload! will catch validation errors and push them into the job error_log

dmlond committed:
uses upload! again to ensure errors are captured in job error_log

dmann committed:
Use default duration for sleep call

dmlond committed:
removed 'has_integrity_exception' boolean attribute from migration
instead, model uses the existence of an error to determine if it has
an integrity exception (no other error states exist at this time)

UploadSpec expects Upload to
 - implement #has_integrity_exception? method
   - returns true if the upload.error_at is not_nil
   - returns false if the upload.error_at is nil
 - mocks has_integrity_exception? call to test whether
   temporary_url returns an IntegrityException or not

UploadSerializerSpec no longer expects UploadSerializer to serialize
has_integrity_exception in the status

RequestSharedExamples mocks has_integrity_exception? call to test whether
an eventually consistent integrity exception endpoint returns an integrity
exception response or not

dmann committed:
Use work_duration for sleep

dmann committed:
Ensure new documents are indexed

dmann committed:
MessageLogWorker#worker_message_limit is not used, went a different direction with implementation

dmlond committed:
fixed expectations for eventually consistent integrity exception

dmann committed:
Ensure threads are stopped before #index_messages returns

dmlond committed:
removed is_consistent from the serialized payloads

dmann committed:
First pass at queue:message_log:index_messages rake task

dmann committed:
Allow env MESSAGE_LOG_WORK_DURATION to set MessageLogQueueHandler#work_duration

dmlond committed:
Merge pull request #1022 from dmann/DDS-886-message_log_queue_handler

DDS-886 message log queue handler

dmlond committed:
Merge remote-tracking branch 'upstream/develop' into DDS-872-eventual_consistency_status
